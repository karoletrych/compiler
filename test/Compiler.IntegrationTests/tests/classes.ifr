module Classes

class Animal 
{
    val _noise : string

    construct (noise : string)
    {
        _noise = noise;
    }

    fun MakeNoise
    {
        System::Console:.WriteLine(_noise);
    }
}

class Dog : Animal
{
    construct : ("Hau!")
    {
    }
}

class Cat : Animal
{
    construct : ("Miau!")
    {
    }
}

class Duck : Animal
{
    construct : ("")
    {
    }

    fun MakeNoise
    {
        System::Console:.WriteLine("Kwak!");
        System::Console:.WriteLine("Kwak!");
        System::Console:.WriteLine("Kwak!");
    }
}

fun createAnimal (animalType : string)
{
    if(animalType == "dog")
        return new Dog();
    else if(animalType == "cat")
        return new Cat();
    else 
        return new Duck();
}

fun printAnimals (animals : System::Collections::Generic::List<Animal>)
{
    var i = 0;
    System::Console:.WriteLine(i);
    System::Console:.WriteLine(animals.Count);
    while(i < animals.Count)
    {
        System::Console:.WriteLine(i);
        val animal = animals.get_Item(i);
        System::Console:.WriteLine(animal.GetType());
        animal.MakeNoise();
        i = i + 1;
    }
}

fun main
{
    val dog = createAnimal("dog");
    val cat = createAnimal("cat");
    val duck = createAnimal("duck");

    val animals = [dog; cat; duck];
    printAnimals(animals);
}
