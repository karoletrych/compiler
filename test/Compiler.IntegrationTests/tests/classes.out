0
3
0
Classes+Dog
Hau!
1
Classes+Cat
Miau!
2
Classes+Duck
Kwak!
Kwak!
Kwak!
